p2p/service/gossip_listener.go

```diff
@@ -0,0 +1,60 @@
+package service
+
+import (
+	"github.com/spacemeshos/go-spacemesh/log"
+	"github.com/spacemeshos/go-spacemesh/priorityq"
+)
+
+type GossipDataHandler func(data GossipMessage, syncer Syncer)
+
+type Listener struct {
+	*log.Log
+	net      Service
+	channels []chan GossipMessage
+	stoppers []chan struct{}
+	syncer   Syncer
+}
+
+func NewListener(net Service, syncer Syncer, log log.Log) *Listener {
+	return &Listener{
+	Log: &log,
+	net: net,
+	syncer:syncer,
+	}
+}
+
+type Syncer interface {
+	FetchPoetProof(poetProofRef []byte) error
+	ListenToGossip() bool
+	IsSynced() bool
+}
+
+func (l *Listener) AddListener(channel string, priority priorityq.Priority, dataHandler GossipDataHandler) {
+	ch := l.net.RegisterGossipProtocol(channel, priority)
+	stop := make(chan struct{})
+	l.channels = append(l.channels, ch)
+	l.stoppers = append(l.stoppers, stop)
+	go l.listenToGossip(dataHandler, ch, stop)
+}
+
+func (l *Listener) Stop() {
+	for _, ch := range l.stoppers {
+		close(ch)
+	}
+}
+
+func (l *Listener) listenToGossip(dataHandler GossipDataHandler, gossipChannel chan GossipMessage, stop chan struct{}) {
+	l.Info("start listening")
+	for {
+		select {
+		case <-stop:
+			return
+		case data := <-gossipChannel:
+			if !l.syncer.ListenToGossip() {
```

Comment:
at least this part could be more generic, (not sure if its needed though). like passing an array of boolean functions and checking them.
though syncer is passed to the data handler so it won't be decoupled eventually..

Reply:
there is no other checks currently, when we have 2+ checks for all listeners we can refactor this
also, keep in mind I didn't refactor block listener yet. this is after I merge this one